test/compiled_metric_test.rb

```diff
@@ -0,0 +1,420 @@
+    # Create a client with default sample rate
+    @sink = StatsD::Instrument::CaptureSink.new(parent: StatsD::Instrument::NullSink.new)
+    client = StatsD::Instrument::Client.new(
+      sink: @sink,
+      prefix: "test",
+      default_tags: [],
+      enable_aggregation: false,
+      default_sample_rate: 1.0,
+    )
+    StatsD.singleton_client = client
+
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+    )
+
+    # Should use client's default sample rate
+    metric.increment(value: 1)
+    assert_equal(1, @sink.datagrams.size)
+  end
+end
+
+class CompiledMetricWithAggregationTest < Minitest::Test
+  def setup
+    @old_client = StatsD.singleton_client
+    @sink = StatsD::Instrument::CaptureSink.new(parent: StatsD::Instrument::NullSink.new)
+    @aggregator = StatsD::Instrument::Aggregator.new(
+      @sink,
+      StatsD::Instrument::DatagramBuilder,
+      "test",
+      [],
+      flush_interval: 0.1,
+    )
+    @client = StatsD::Instrument::Client.new(
+      sink: @sink,
+      prefix: "test",
+      default_tags: [],
+      enable_aggregation: true,
+    )
+    @client.instance_variable_set(:@aggregator, @aggregator)
+    StatsD.singleton_client = @client
+  end
+
+  def teardown
+    @sink.clear
+    StatsD.singleton_client = @old_client
+  end
+
+  def test_aggregates_precompiled_metrics
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      tags: { shop_id: Integer },
+    )
+
+    metric.increment(shop_id: 123, value: 1)
+    metric.increment(shop_id: 123, value: 2)
+    metric.increment(shop_id: 123, value: 3)
+
+    @aggregator.flush
+
+    assert_equal(1, @sink.datagrams.size)
+    datagram = @sink.datagrams.first
+    assert_equal(6, datagram.value) # 1 + 2 + 3
+    assert_equal(["shop_id:123"], datagram.tags)
+  end
+
+  def test_aggregates_different_tag_combinations_separately
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      tags: { shop_id: Integer },
+    )
+
+    metric.increment(shop_id: 123, value: 1)
+    metric.increment(shop_id: 456, value: 2)
+    metric.increment(shop_id: 123, value: 3)
+
+    @aggregator.flush
+
+    assert_equal(2, @sink.datagrams.size)
+
+    shop_123_datagram = @sink.datagrams.find { |d| d.tags.include?("shop_id:123") }
+    shop_456_datagram = @sink.datagrams.find { |d| d.tags.include?("shop_id:456") }
+
+    assert_equal(4, shop_123_datagram.value) # 1 + 3
+    assert_equal(2, shop_456_datagram.value)
+  end
+
+  def test_aggregates_static_tag_metrics
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      static_tags: { service: "web" },
+    )
+
+    metric.increment(value: 1)
+    metric.increment(value: 2)
+    metric.increment(value: 5)
+
+    @aggregator.flush
+
+    assert_equal(1, @sink.datagrams.size)
+    datagram = @sink.datagrams.first
+    assert_equal(8, datagram.value) # 1 + 2 + 5
+  end
```

Comment:
Can you add a test for allocations please? The aggregated codepath should have 0 allocations

Reply:
Yup. I can move that one here too 👍🏼

Reply:
https://github.com/Shopify/statsd-instrument/pull/401/commits/13fcdae7c62c8a0bf5346c1b2ab0bfd86525ef06 added. However, I noticed that I get 2 allocations instead of 0 when I compare this to SFR. We're using two different methods for measuring allocations, so I think that might explain the discrepancy. Regardless, it's still a significant improvement compared to the 145 allocations I was seeing when running against the old `StatsD.increment`. Additionally, I had to disable this for JRuby since it apparently doesn't support GC.stats.